### PR TITLE
network: Display DHCPv6 DUID

### PR DESCRIPTION
**Pull request by ssahani**
```
~/tt/dhcpv6-duid dhcpv6-duid* 12s                                                                                                                                             3h5m ⚑ ◒  
❯ build/networkctl status veth99
● 8: veth99                                                                    
                     Link File: /usr/lib/systemd/network/99-default.link       
                  Network File: /usr/lib/systemd/network/veth99.network        
                          Type: ether                                          
                         State: routable (configured)
                        Driver: veth                                           
                    HW Address: 1e:04:f8:b8:2f:d4                              
                           MTU: 1500 (min: 68, max: 65535)                     
                         QDisc: noqueue                                        
  IPv6 Address Generation Mode: eui64                                          
          Queue Length (Tx/Rx): 1/1                                            
              Auto negotiation: no                                             
                         Speed: 10Gbps                                         
                        Duplex: full                                           
                          Port: tp                                             
                       Address: 192.168.5.56 (DHCP4)                           
                                2600::10                                       
                                fe80::1c04:f8ff:feb8:2fd4                      
                       Gateway: 192.168.5.1                                    
                                fe80::b459:12ff:fec8:8193                      
               DHCP4 Client ID: IAID:0xd0cc9409/DUID                           
             DHCP6 Client IAID: 0xd0cc9409                                     
             DHCP6 Client DUID: DUID-EN/Vendor:0000ab11d258482fc7eee6510000 
```

### CODE REVIEW

File: src/libsystemd-network/sd-dhcp6-client.c [hunk 1]
```diff
@@ -331,6 +332,48 @@ int sd_dhcp6_client_set_duid_llt(
...
+
+        v = dhcp6_duid_type_to_string(be16toh(client->duid.type));
+        if (v) {
+                s = strdup(v);
+                if (!s)
+                        return -ENOMEM;
+        } else {
+                r = asprintf(&s, "%0x", client->duid.type);
+                if (r < 0)
+                        return r;
```
Comment (poettering) [inline src/libsystemd-network/sd-dhcp6-client.c]: asprintf() returns -1 on error, i.e. this must be `return -ENOMEM`, not `return r`;

File: src/libsystemd-network/sd-dhcp6-client.c [hunk 2]
```diff
@@ -331,6 +332,48 @@ int sd_dhcp6_client_set_duid_llt(
...
+        }
+
+        t = hexmem(&client->duid.raw.data, client->duid_len);
+        if (!t)
+                return -ENOMEM;
+
+        p = strjoin(s, ":", t);
+        if (!p)
+                return ENOMEM;
```
Comment (poettering) [inline src/libsystemd-network/sd-dhcp6-client.c]: There's a `-` missing here

File: src/network/networkctl.c [hunk 3]
```diff
@@ -2089,6 +2089,14 @@ static int link_status_one(
                                    TABLE_EMPTY,
                                    TABLE_STRING, "DHCP6 Client IAID:",
                                    TABLE_STRING, iaid);
+        }
```
Comment (poettering) [inline src/network/networkctl.c:2094]: error checking got lost here

File: src/libsystemd-network/sd-dhcp6-client.c (hunk 2 above)
Comment (ssahani) [reply src/libsystemd-network/sd-dhcp6-client.c]: Ouch

File: src/libsystemd-network/sd-dhcp6-client.c [hunk 4]
```diff
@@ -331,6 +332,48 @@ int sd_dhcp6_client_set_duid_llt(
...
 
+static const char* const dhcp6_duid_type_table[_DUID_TYPE_MAX] = {
+        [DUID_TYPE_LLT]  = "DUID-LLT",
+        [DUID_TYPE_EN]   = "DUID-EN/Vendor",
+        [DUID_TYPE_LL]   = "DUID-LL",
+        [DUID_TYPE_UUID] = "UUID",
+};
+DEFINE_PRIVATE_STRING_TABLE_LOOKUP_TO_STRING(dhcp6_duid_type, DUIDType);
+
+int sd_dhcp6_client_duid_to_string(
```
Comment (poettering) [inline src/libsystemd-network/sd-dhcp6-client.c]: hmm, I figure `_as_string` would be better than `_to_string` here I guess

File: src/network/networkd-link.c [hunk 5]
```diff
@@ -4084,7 +4084,7 @@ static void serialize_addresses(
 
 int link_save(Link *link) {
         const char *admin_state, *oper_state, *carrier_state, *address_state;
-        _cleanup_free_ char *temp_path = NULL;
+        _cleanup_free_ char *temp_path = NULL, *duid = NULL;
```
Comment (poettering) [inline src/network/networkd-link.c]: move this new variable declaration into the innermost scope where it's needed? And similar for iaid?

File: src/libsystemd-network/sd-dhcp6-client.c (hunk 4 above)
Comment (ssahani) [reply src/libsystemd-network/sd-dhcp6-client.c]: OK done.

**Pull request merged**